compiler/frontend/src/hir_to_mir.rs

```diff
@@ -600,6 +584,24 @@ impl<'a> LoweringContext<'a> {
     }
 }
 
+#[extension_trait]
+impl IsExact for hir::Pattern {
+    fn is_exact(&self) -> bool {
+        match self {
+            Self::NewIdentifier(_) => false,
+            Self::Int(_) => true,
+            Self::Text(_) => true,
+            Self::Tag { symbol: _, value } => value.as_ref().map_or(true, |val| val.is_exact()),
+            Self::List(items) => items.iter().all(IsExact::is_exact),
+            Self::Struct(struct_) => struct_
+                .iter()
+                .all(|field| field.0.is_exact() && field.1.is_exact()),
```

Comment:
Structs are more lenient when matching. With this change, we can also merge `compile_pattern(…)` and `compile_key_pattern(…)`

```suggestion
            Self::Struct(struct_) => false,
```